### ISSUE TITLE
igraph_betweenness_subset() invalid result when 

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#2359]
**Describe the bug**

`igraph_betweenness_subset()` returns incorrect results when the graph is undirected an the source and target sets differ.

The underlying reason is that the implementation divides the shortest path count by 2 in the undirected case, making the assumption that for any undirected `a -> b` path there is an equivalent `b -> a` path. However, this is only true if both of `a` and `b` are both targets and sources.

**To reproduce**

In this graph, the subset betweenness of vertex `1` with sources of {0, 1} and targets of {2, 3} is 1:

<img width="360" alt="image" src="https://github.com/igraph/igraph/assets/1212871/d81ccae7-cdab-4653-a2ef-9ba820ce20ef">

There is one shortest path between these two subsets that passes through vertex `1`.

However, igraph returns 0.5. In Python:

```
In [1]: import igraph as ig

In [2]: g = ig.Graph([(0,1), (1,2)], n=4)

In [3]: g.betweenness(sources=[0,1], targets=[2,3])
Out[3]: [0.0, 0.5, 0.0, 0.0]
```

**Version information**
Current `master`, e84149a1cc01d9797dad753780d87dfdadbd6885

**Comment by szhorvat** [comment]
An interesting observation here is the following: 

With the usual betweenness, we get equivalent results from an undirected graph and its directed analog where each undirected edge is replaced by reciprocal directed ones. More precisely, the directed result is twice the undirected result.

This equivalency breaks down with subset betweenness when the source and target sets are neither the same nor disjoint, but _partially_ overlapping.

**Comment by szhorvat** [comment]
NetworkX suffers from the same issue, and this is documented, although in a somewhat confusing manner:

https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.centrality.betweenness_centrality_subset.html

**Comment by ntamas** [comment]
Hmm, okay, what shall the fix be then? Am I right to think that if there are _u_ sources in $S$, _v_ targets in $T$ and an intersection size of _x_, then paths from $S \cap T$ to $S \cap T$ are counted twice, but paths from $S \setminus T$ to $T$ or from $S$ to $T \setminus S$ are only counted once?

**Comment by ntamas** [comment]
As a stopgap measure, maybe we should handle the case of full overlap and no overlap properly and print a warning for partial overlaps until we work the math out?

**Comment by szhorvat** [comment]
> As a stopgap measure, maybe we should handle the case of full overlap and no overlap properly and print a warning for partial overlaps until we work the math out?

Yes, exactly. I expect that the majority of uses cases will be for zero overlap.

**Comment by ntamas** [comment]
Maybe [this one](https://snap.stanford.edu/class/cs224w-readings/brandes01centrality.pdf)?

Also, two more things:

1. I'm not sure whether the current imlpementation works correctly if a vertex appears more than once in the source selector. This needs to be checked. (Targets appearing multiple times is okay because we use a `bool` array to keep track of which vertices are targets).
2. Testing the size of the overlap between `sources` and `targets` is not trivial. If we expect both the source and target set to be small compared to the size of the graph, we can probably convert them to vectors, sort them and take the intersection with `igraph_vector_intersect_sorted()` and this will be negligible compared to all the other stuff we need to do. However, if `sources` or `targets` covers all vertices, this is a waste of time. Maybe we should add internal `igraph_i_are_vertex_selectors_disjoint()` and `igraph_i_are_vertex_selectors_covering_the_same_vertices()` helper function that could special-case the "all" and "none" vertex selectors at least? (I don't like the name of the second function but we cannot use `are_vertex_selectors_same()` because there could be vertex selectors that are _technically_ different but happen to yield the same set of vertices).

**Comment by szhorvat** [comment]
A quick solution here would be to split the computation into an overlapping and a non-overlapping part when the graph is undirected.

First we run it with $S \cap T$ as both the source and target, and scale the result by $\frac{1}{2}$. Then we run it with $S \setminus T$ as the source and $T \setminus S$ as the target and leave the result as-is.

Finally we add up the two results.

There may be more efficient ways than this, but currently this seem like the least effort and it fixes the issue.
